risc0/povw/src/lib.rs

```diff
@@ -0,0 +1,996 @@
+        let mut path = Box::<Array<Digest, _>>::default();
+        index >>= 8;
+        for i in 0..path.len() {
+            path[i] = self.subtree_root(i, index ^ U256::from(1));
+            index >>= 1;
+        }
+
+        Opening { bitmap, path }
+    }
+
+    fn subtree_root(&self, height: usize, index: U256) -> Digest {
+        assert!(height <= Self::TREE_HEIGHT);
+        assert!(index < (U256::from(1) << (Self::TREE_HEIGHT - height)));
+
+        if height < WorkLog::TREE_HEIGHT {
+            let sublog_index_bits = WorkLog::TREE_HEIGHT - height;
+            let sublog_index =
+                (index & ((U256::from(1) << sublog_index_bits) - U256::from(1))).to::<U96>();
+            let log_index = (index >> sublog_index_bits).to::<U160>();
+            return self
+                .logs
+                .get(&log_index)
+                .unwrap_or(&WorkLog::EMPTY)
+                .subtree_root(height, sublog_index);
+        }
+
+        // Cast index to U160. At this point it must be less than 2^160.
+        let index: U160 = index.to();
+        let height_offset = height - WorkLog::TREE_HEIGHT;
+
+        // A level of the tree, holding only roots of non-empty subtrees that are decendents of the
+        // desired root.
+        let mut level: BTreeMap<U160, Digest> = self
+            .logs
+            .iter()
+            .filter(|(id, _)| id.wrapping_shr(height_offset) == index)
+            .map(|(id, job)| (*id, job.commit()))
+            .collect();
+
+        // Index is of an empty subtree
+        if level.is_empty() {
+            return EMPTY_SUBTREE_ROOTS[height];
+        }
+
+        // Iterate up the levels, from the roots of the job trees to the root of the subtree.
+        #[allow(clippy::needless_range_loop)]
+        for i in WorkLog::TREE_HEIGHT..height {
+            let next_level_keys: BTreeSet<U160> =
+                level.keys().map(|idx| idx.wrapping_shr(1)).collect();
+            level = next_level_keys
+                .into_iter()
+                .map(|idx| {
+                    let empty = &EMPTY_SUBTREE_ROOTS[i];
+                    let left = level.get(&(idx << 1)).unwrap_or(empty);
+                    let right = level.get(&((idx << 1) ^ U160::from(1))).unwrap_or(empty);
+                    (idx, join(*left, *right))
+                })
+                .collect();
+        }
+
+        // At this point, there should be exactly one node in the level.
+        assert!(level.len() == 1);
+        *level.first_key_value().unwrap().1
+    }
+
+    fn bitmap_at(&self, index: U256) -> Bitmap {
+        self.logs
+            .get(&(index >> 96usize).to::<U160>())
+            .unwrap_or(&WorkLog::EMPTY)
+            .bitmap_at(index.wrapping_to::<U96>())
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct WorkLog {
+    pub jobs: BTreeMap<u64, Job>,
+}
+
+impl WorkLog {
+    pub const TREE_HEIGHT: usize = Job::TREE_HEIGHT + 64;
```

Comment:
Why 64?

Reply:
I've done some work to try to make this more clear in 25c5694a2daae335b09f65a1a98b2a684c324352